OpenRA.Game/GameRules/ActorInfo.cs

```diff
@@ -115,27 +115,34 @@ public IEnumerable<TraitInfo> TraitsInConstructOrder()
 			}).ToList();
 
 			var resolved = source.Where(s => s.Dependencies.Count == 0 && s.OptionalDependencies.Count == 0).ToList();
-			var unresolved = source.Except(resolved);
+			var unresolved = source.ToHashSet();
+			unresolved.ExceptWith(resolved);
 
-			var testResolve = new Func<Type, Type, bool>((a, b) => a == b || a.IsAssignableFrom(b));
+			bool TestResolve(Type a, Type b) => a.IsAssignableFrom(b);
 
 			// This query detects which unresolved traits can be immediately resolved as all their direct dependencies are met.
 			var more = unresolved.Where(u =>
 				u.Dependencies.All(d => // To be resolvable, all dependencies must be satisfied according to the following conditions:
-					resolved.Exists(r => testResolve(d, r.Type)) && // There must exist a resolved trait that meets the dependency.
-					!unresolved.Any(u1 => testResolve(d, u1.Type))) && // All matching traits that meet this dependency must be resolved first.
+					resolved.Exists(r => TestResolve(d, r.Type)) && // There must exist a resolved trait that meets the dependency.
+					!unresolved.Any(u1 => TestResolve(d, u1.Type))) && // All matching traits that meet this dependency must be resolved first.
 				u.OptionalDependencies.All(d => // To be resolvable, all optional dependencies must be satisfied according to the following condition:
-					!unresolved.Any(u1 => testResolve(d, u1.Type)))); // All matching traits that meet this optional dependencies must be resolved first.
+					!unresolved.Any(u1 => TestResolve(d, u1.Type)))); // All matching traits that meet this optional dependencies must be resolved first.
 
 			// Continue resolving traits as long as possible.
 			// Each time we resolve some traits, this means dependencies for other traits may then be possible to satisfy in the next pass.
-			while (more.Any())
-				resolved.AddRange(more);
+			var readyToResolve = more.ToList();
+			while (readyToResolve.Any())
```

Comment:
```suggestion
			while (readyToResolve.Count != 0)
```

We had chosen to use this style

Reply:
Updated